Proposal for Extracting Load Balancing from the Event Processor

christothes opened the pull request:
This PR is for discussion of #9051 and demonstrates the key proposed functional refactoring. It keeps all existing tests in tact and doesn't address details such as missing comments. 

## Summary
There are two primary concerns owned by the current `EventProcesorClient` implementation

1. Managing the processing tasks for each partition owned
1. Managing ownership of partitions
   - A subset of which is Load balancing of ownership

This PR proposes that `EventProcesorClient` can offload all of concern 2 into the new `PartitionLoadBalancer` class.

### Changes of Interest
- All code dealing with the acquisition and relinquishing of partition ownership moves to `PartitionLoadBalancer`. This means the following properties and methods migrate over.
  - `RandomNumberGenerator` and `s_randomSeed `
  - `InstanceOwnership` (see details below)
  - `ActiveOwnershipWithDistribution`
  - `FindAndClaimOwnershipAsync`
  - `ClaimOwnershipAsync`
  - `RenewOwnershipAsync`
  - `RelinquishOwnershipAsync`
 - `InstanceOwnership` is now encapsulated inside `PartitionLoadBalancer` as it is mostly no longer a concern of `EventProcesorClient`. The one exception is the requirement of getting the list of currently owned partitionIds to prune running processor tasks. For this scenario we expose a new property on  `PartitionLoadBalancer` called `OwnedPartitionIds`
 - Error handling newly residing in `PartitionLoadBalancer` previously serviced by calls to `OnProcessErrorAsync` are now bubbled up as a `EventHubsException` with the operation string passed as the exception message. See 4aa3a8fce0c32e77197721c6a56a94888b0e58f1 for an alternate approach using an injected `OnProcessErrorAsync`. Ultimately I concluded that passing the error handler is more error prone and leaks implementation details across class boundaries.

note: This PR is a comparison to #8816 until it is merged to master

christothes committed:
wip - switching machines

JoshLove-msft committed:
reliability fixes (#9095)

JoshLove-msft committed:
don't throw if table info included in conn string (#9114)

christothes committed:
changes for discussions

mikeharder committed:
[Storage] Disable live test parallelism and increase timeout (#9136)

- Add template parameters to allow individual pipelines to control parallel assembly execution and timeout.
- Storage live tests have been unreliable, and running multiple test assemblies in parallel may be contributing.
- Storage live tests have also been timing out at 3 hours, increasing to 5 hours allows tests to finish and give more useful results.
- Once tests are stable, parallel should be re-enabled (#9144) and timeout should be decreased (#9143).

mikeharder committed:
[Storage] Set live test timeout to 20 minutes (#9137)

- Add LiveOnly attribute to Assert.Inconclusive() tests
  - Workaround for NUnit bug "TimeoutAttribute makes Assert.Inconclusive() count as failure" (nunit/nunit#3282)

christothes committed:
tweaks to error handling

heaths committed:
Update Certificates API for cross-language consistency (#9123)

* Update Certificates API for cross-language consistency

Fixes #9102

* Update saved Certificates API

* Resolve review meeting feedback

* Update saved Certificates API

jsquire committed:
[Event Hubs Client] Track Two (Processor Event Members Tweak) (#9115)

The focus of these changes is to re-scope visibility for the
On[[ EVENT]] methods used for raising events within the processor.  Previously,
these followed the expected .NET event pattern of being protected/virtual.

After discission, it was decided that these should not be exposed, as we
do not have intended scenarios around extending the processor client.

jsquire committed:
[Event Hubs Client] Track Two (Telemetry Name Casing) (#9118)

The focus of these changes is to normlize the names for telemetry properties
to lower-case, to match the format used across other langugages and in the
track one client.

laurawu19 committed:
Support adding DiskEncryptionSet in existing disk (#9094)

* Support adding DiskEncryptionSet in existing disk

* Regenerate sdk with merged swagger and update tests

* Remove private branch info

AlexanderSher committed:
Fix #7248: Address ConnectionString parsing in AppConfiguration (#9138)

* Fix #7248: Address ConnectionString parsing in AppConfiguration

* Address PR comments

* Delete unused file.

* - Remove redundant features
- Change messages

* Address PR comment

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/EventProcessorClient.cs:
```diff
@@ -261,6 +255,8 @@ public bool IsRunning
         ///
         private PartitionManager StorageManager { get; }
 
+        private PartitionLoadBalancer LoadBalancer { get; }
```
// TODO:  Appease the pedantic bald reviewer with doc comments.   😛

christothes replied:
:smile:

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Shared/src/Processor/PartitionLoadBalancer.cs:
```diff
@@ -0,0 +1,439 @@
+        ///   The set of partition ownership this event processor owns.  Partition ids are used as keys.
+        /// </summary>
+        ///
+
```
nit:  Extra blank line

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Shared/src/Processor/PartitionLoadBalancer.cs:
```diff
@@ -0,0 +1,439 @@
+        /// <param name="partitionIds"></param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        ///
+        internal virtual async ValueTask<PartitionOwnership> RunAsync(string[] partitionIds, CancellationToken cancellationToken)
```
May as well designate all `internal` members public, as the class visibility constrains.   Not that it is impactful, but that's the pattern that has been used.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Shared/src/Processor/PartitionLoadBalancer.cs:
```diff
@@ -0,0 +1,439 @@
+            EventHubName = eventHubName;
+            ConsumerGroup = consumerGroup;
+            OwnershipExpiration = ownershipExpiration;
+
```
nit: Extra blank line

jsquire commented:
I think the high level makes sense.   There's some naming that we'll probably want to talk about.     The error handling is an interesting twist.   Call me out if I'm missing something but, If I'm reading correctly, each call to `RunAsync` (which we may want to consider renaming to something like `PerformBalancing` or `ReconsiderOwnership` or something) would be the place where the exception would bubble for the critical areas.   The utility methods would also bubble their own.     

I *think* that covers us, as we'd have a deterministic spot for errors to surface that wouldn't result in log-n-forget or wait-until-stopped.    Assuming that's the case, I agree with you in that I like that trade-off better than injecting a handler or having another event-style callback.

jsquire commented on sdk/eventhub/Azure.Messaging.EventHubs.Shared/src/Processor/PartitionLoadBalancer.cs:
```diff
@@ -0,0 +1,439 @@
+        ///   Interacts with the storage system with responsibility for creation of checkpoints and for ownership claim.
+        /// </summary>
+        ///
+        private readonly PartitionManager StorageManager;
```
We'd have to move `PartitionManager` into the shared area as well, which we 're likely to need regardless.   I'm guessing `Shared` isn't able to compile in the current state.

christothes replied:
As I recall, Shared is a shared project (shproj) - does it compile on its own? 
The Processor project builds and runs tests successfully though.

christothes replied:
Oh, I see now if I build the .sln

jsquire replied:
Yeah, the goal is for the shared code to be fully stand-alone, including responsibility for managing its own tests via the associated test project.   When one of the other projects pulls in a bundle, it should have confidence that the code is stable.  

Because the test project pulls in all of the bundles from the shared code, anything in there needs to be compilable.   Since we'd want to have load balancing testable, we'd need to pull the `PartitionManager` in for that anyway.

annelo-msft committed:
Add docstrings to TextAnalytics public types. (#9153)

* Slight renaming and adding docstrings

* adding more docstrings

* remove empty docstrings on internal methods

* fix tests per refactoring

* pr feedback

* pr feedback

* pr feedback

* revert readme changes

* readme verifications

* actually revert readme

christothes commented:
> I think the high level makes sense. There's some naming that we'll probably want to talk about. The error handling is an interesting twist. Call me out if I'm missing something but, If I'm reading correctly, each call to `RunAsync` (which we may want to consider renaming to something like `PerformBalancing` or `ReconsiderOwnership` or something) would be the place where the exception would bubble for the critical areas. The utility methods would also bubble their own.
> 
> I _think_ that covers us, as we'd have a deterministic spot for errors to surface that wouldn't result in log-n-forget or wait-until-stopped. Assuming that's the case, I agree with you in that I like that trade-off better than injecting a handler or having another event-style callback.

That's correct. This approach lets the Processor control how those errors translate to event handlers and leaves the LoadBalancer agnostic. Effectively the handling logic still resides in `RunAsync`

I was on the fence regarding the naming. Either we could go more specific at the method level, as you suggested, or let the class provide the context as to what `RunAsync` means, Arguably if all the `PartitionLoadBalancer` does is run or stop, it's implied that they are starting and stopping load balancing. However, I don't feel strongly either way.

christothes commented:
@jsquire Thanks for the comments. Once you confirm you are good with the general approach, I'll button up the details and refactor the tests.

chidozieononiwu committed:
Prepare Repo for DocWarden Updates (#9004)

* Exclude CodeGenerationLibrary from package indexing exclusion list

* Update Packages.md

* Rename Readmes for consistency

* Rename Changelogs for Consistency

* Update Changelogs for Storage

* Update Changelogs

* Add latest release notes for various packages

* Add titles to release notes

* Modify changelogs to match new guidiance

* Add Missing CHANGELOG.md for Text Analytics

* Update Changelog Format

JoshLove-msft committed:
Avoid GUID collisions (#9140)

jsquire reviewed (commented):
> @jsquire Thanks for the comments. Once you confirm you are good with the general approach, I'll button up the details and refactor the tests.

I'm good with the approach.   With the confirmation on the error behavior, I'm in agreement about the trade-off there.

jsquire commented:
> I was on the fence regarding the naming. Either we could go more specific at the method level, as you suggested, or let the class provide the context as to what `RunAsync` means, Arguably if all the `PartitionLoadBalancer` does is run or stop, it's implied that they are starting and stopping load balancing. However, I don't feel strongly either way.

That's a fair point.    I worry that `RunAsync` may be confusing in the context of the processor where that has some history as the "thing to call for running the background stuff."   Not sure that we need to go all out, but "Start/Stop" would also seem to imply semantics about things actively running in the background.   Something that gives a sense of "We're going to run one phase/pass/whatever of load balancing in response to this call" is kind of what I was aiming for.

We probably need to think about the "Stop" call the same way.   We're kind of saying "Finalize load balancing for this" more than "Stop doing the thing in the background".

christothes commented:
> > I was on the fence regarding the naming. Either we could go more specific at the method level, as you suggested, or let the class provide the context as to what `RunAsync` means, Arguably if all the `PartitionLoadBalancer` does is run or stop, it's implied that they are starting and stopping load balancing. However, I don't feel strongly either way.
> 
> That's a fair point. I worry that `RunAsync` may be confusing in the context of the processor where that has some history as the "thing to call for running the background stuff." Not sure that we need to go all out, but "Start/Stop" would also seem to imply semantics about things actively running in the background. Something that gives a sense of "We're going to run one phase/pass/whatever of load balancing in response to this call" is kind of what I was aiming for.
> 
> We probably need to think about the "Stop" call the same way. We're kind of saying "Finalize load balancing for this" more than "Stop doing the thing in the background".

I see your point, and I think you convinced me that we need some more specific terminology, I'll think a bit more about it and propose something in the next commit.

chidozieononiwu committed:
Fix Errors in changelogs (#9168)

christothes committed:
project organization and PartitionLoadBalancerTests

christothes committed:
Refactor Event Processor Client load balancing algorithm to improve performance (#8816)

Updates to the load balancing implementation used by the Event Processor client.

jsquire committed:
[Event Hubs Client] Track Two (Fix Processor Resume from Checkpoint) (#9166)

The focus of these changes is to correct an issue where the Event Processor
client would resume from a checkpoint using the event that the checkpoint
was based on, resulting in that event being processed multiple times.

The proccessor should assume the checkpoint is based on the event last processed
and should, therefore, resume from the next event available in the stream.

jsquire committed:
[Event Hubs Client] Track Two (Timer Dispose Race Condition Fix) (#9156)

The focus of these changes is to address a race condition that could
cause the authorization refresh timer callback to throw an ObjectDisposedException
due to a set of benign race conditions.  These changes attempt to make those
conditions less likely but cannot elimitate them completely without introducing
synchronization overhead.  Because the condition is benign, accept the
resulting exception as an expected case rather than paying that cost.

ShivangiReja committed:
Add Migration Guide for Event Hubs (#9171)

Add Migration Guide for Event Hubs

christothes committed:
converted EventProcessorClientTests

christothes committed:
merge to upstream

christothes commented:
closing this PR in favor of #9198
